### PR TITLE
Trade supports multiple routes (AggTrade)

### PR DESCRIPTION
**Pull request by willpote**
Same as https://github.com/Uniswap/uniswap-v3-sdk/pull/55 but with a new AggregatedTrade object (avoids major version bump)

### CODE REVIEW

File: src/swapRouter.ts [hunk 1]
```diff
@@ -74,34 +75,32 @@ export abstract class SwapRouter extends SelfPermit {
...
-    trades: Trade<Currency, Currency, TradeType> | Trade<Currency, Currency, TradeType>[],
+    trade: Trade<Currency, Currency, TradeType> | AggregatedTrade<Currency, Currency, TradeType>,
     options: SwapOptions
   ): MethodParameters {
-    if (!Array.isArray(trades)) {
-      trades = [trades]
+    const isAggregated = (
+      trade: Trade<Currency, Currency, TradeType> | AggregatedTrade<Currency, Currency, TradeType>
+    ): trade is AggregatedTrade<Currency, Currency, TradeType> => {
+      return (trade as AggregatedTrade<Currency, Currency, TradeType>).trades !== undefined
```
Comment (moodysalem) [inline src/swapRouter.ts:84]: why not do instance of, and also move to an external fn

File: src/swapRouter.ts [hunk 2]
```diff
@@ -74,34 +75,32 @@ export abstract class SwapRouter extends SelfPermit {
...
-    // All trades should have the same starting and ending token.
-    invariant(
-      trades.every(trade => trade.route.tokenPath[0].equals(tokenIn)),
-      'TOKEN_IN_DIFF'
-    )
-    invariant(
-      trades.every(trade => trade.route.tokenPath[trade.route.tokenPath.length - 1].equals(tokenOut)),
-      'TOKEN_OUT_DIFF'
-    )
+    // All trades have the same starting and ending token (the trade objects validate this)
```
Comment (moodysalem) [inline src/swapRouter.ts:94]: seems no longer checked

File: src/entities/trade-aggregated.ts [hunk 3]
```diff
@@ -0,0 +1,223 @@
...
+   * @param tradeType The type of trade, exact input or exact output
+   */
+  constructor({
+    trades,
+    inputAmount,
+    outputAmount,
+    tradeType
+  }: {
+    trades: Trade<TInput, TOutput, TTradeType>[]
+    inputAmount: CurrencyAmount<TInput>
```
Comment (moodysalem) [inline src/entities/trade-aggregated.ts:131]: does not need to be specified, can just be a lazy getter that computes it from the trades

**Comment by willpote** [comment]
Discussed offline, Moody suggested a way to do the other version of this PR in a backward compataible way. Closing this in favour of: https://github.com/Uniswap/uniswap-v3-sdk/pull/55/files

**Pull request closed without merging**